### PR TITLE
Azure Service Discovery resource group filter

### PR DESCRIPTION
**Pull request by David-N-Perkins**
Added an optional `resource_group` configuration for Azure service discovery, which limits the discovery when specified. See issue [10253](https://github.com/prometheus/prometheus/issues/10253).

Also updated a deprecated azure auth call.

### CODE REVIEW

**Comment by David-N-Perkins** [comment]
@roidelapluie I implemented this using a new config value. Let me know if there's another preferred method to get the resource group.

**Comment by roidelapluie** [comment]
Thanks for this! It seems like we would need azure users to test this.

**Comment by jrauschenbusch** [comment]
@David-N-Perkins Thanks for implementing this. A new optional config value seems to be the best solution for this topic. It's backwards compatible and provide a way to reduce the amount of ARM calls massively if it's a subscription containing a lot of VMs. I'll try out you code changes and give you feedback. But maybe there are other Azure users who can also give feedback. Especially the MSI auth change will be hard to test locally.

**Comment by jrauschenbusch** [comment]
@David-N-Perkins Sorry for the late response. I've now tested your branch for ~24h within an AKS cluster.

**TL;DR**: LGTM! @roidelapluie Please merge 😄
- MSI authentication works (tested with azure pod identity in AKS)
- Applying the resource groups parameter also works like a charm (ARM request report comparison (#calls before/#calls after) using [`Export-AzLogAnalyticRequestRateByInterval`](https://docs.microsoft.com/en-us/powershell/module/az.compute/export-azloganalyticrequestratebyinterval?view=azps-7.3.2))
- Backwards compatibility is given by omitting the `resource_group` config parameter (also tested via the report approach mentioned above)

I had previously accidentally built a container image from the main branch and was wondering why the parameter shouldn't exist in `azure.plain` (see log output below). Then I realized it was the wrong branch I was building the image on. But it was a good test that the parameter is accepted as soon as i was working on the correct branch.

Before using the right branch there was the following log output:
```
 ts=2022-03-23T09:59:23.929Z caller=main.go:870 level=error msg="Error reloading config" err="couldn't load configuration (--config.file=\"/etc/config/prometheus.yml\"): parsing YAML file /etc/config/prometheus.yml: yaml: unmarshal errors:\n  line XXX: field resource_group not found in type azure.plain" 
```

After using the right branch everything worked as expected.

Good job! 🚀

**Comment by roidelapluie** [comment]
But is this a beaking change for users? Do they need to change their credentials ?

**Comment by jrauschenbusch** [comment]
No, it's not a breaking change. I've just tested if the MSI authentication still works with the changes made in this PR. Service Principal authentication was not touched and should work as before:

https://github.com/prometheus/prometheus/blob/10b6d005291b51c092a1eac428f50e00290c484b/discovery/azure/azure.go#L201

Just the MSI authentication was replaced, as the current methods are deprecated:

https://github.com/Azure/go-autorest/blob/b3899c1057425994796c92293e931f334af63b4e/autorest/adal/token.go#L738

Furthermore one can omit the resource_group parameter and then the behavior is the same as before. Hence also not a breaking change.

File: discovery/azure/azure.go [hunk 1]
```diff
@@ -429,9 +431,25 @@ func (client *azureClient) getVMs(ctx context.Context) ([]virtualMachine, error)
...
+}
+
+func (client *azureClient) getScaleSets(ctx context.Context, resourceGroup string) ([]compute.VirtualMachineScaleSet, error) {
 	var scaleSets []compute.VirtualMachineScaleSet
-	result, err := client.vmss.ListAll(ctx)
+	var result VmssListResultPage
+	var err error
+	if len(resourceGroup) == 0 {
+		var rtn compute.VirtualMachineScaleSetListWithLinkResultPage
+		rtn, err = client.vmss.ListAll(ctx)
```
Comment (roidelapluie) [inline discovery/azure/azure.go:446]: We should check the error here to avoid deferring a nil pointer

File: discovery/azure/azure.go [hunk 2]
```diff
@@ -429,9 +431,25 @@ func (client *azureClient) getVMs(ctx context.Context) ([]virtualMachine, error)
...
+		result = &rtn
+	} else {
+		var rtn compute.VirtualMachineScaleSetListResultPage
+		rtn, err = client.vmss.List(ctx, resourceGroup)
```
Comment (roidelapluie) [inline discovery/azure/azure.go:450]: Same

**Comment by roidelapluie** [comment]
thanks. I have spin up azure vm's to test this, it indeed works. Added a comment.

File: discovery/azure/azure.go (hunk 1 above)
Comment (David-N-Perkins) [reply discovery/azure/azure.go:446]: Fixed.

**Comment by roidelapluie** [comment]
Thanks!
